src/checkers/inference/model/ConstraintManager.java

```diff
@@ -37,171 +30,186 @@
 
     private VisitorState visitorState;
 
-    private ConstraintVerifier constraintVerifier;
-
     public void init(InferenceAnnotatedTypeFactory inferenceTypeFactory) {
         this.inferenceTypeFactory = inferenceTypeFactory;
         this.realQualHierarchy = inferenceTypeFactory.getRealQualifierHierarchy();
         this.visitorState = inferenceTypeFactory.getVisitorState();
         this.checker = inferenceTypeFactory.getContext().getChecker();
-        this.constraintVerifier = new ConstraintVerifier(realQualHierarchy);
     }
 
     public Set<Constraint> getConstraints() {
         return constraints;
     }
 
-    public ConstraintVerifier getConstraintVerifier() {
-        return constraintVerifier;
-    }
-
-    private void add(Constraint constraint) {
-        if (!ignoreConstraints) {
+    /**
+     * Checks to see if the given {@link Constraint} is an instance of
+     * {@link AlwaysFalseConstraint}. If so, a warning is issued. If not, it adds the
+     * {@link Constraint} to the constraint set only if the {@code ignoreConstraints} flag is set to
+     * false and the constraint is not an {@link AlwaysTrueConstraint}.
+     *
+     * <p>
+     * The source constraint class is used to construct an error key for issue the warning.
+     *
+     * @param constraint a (possibly normalized) constraint
+     * @param nonNormalizedConstraintClass the class literal of the original non-normalized
+     *        constraint
+     */
+    private void add(Constraint constraint,
+            Class<? extends Constraint> nonNormalizedConstraintClass) {
+        if (constraint instanceof AlwaysFalseConstraint) {
```

Comment:
Why doesn't care about `ignoreConstraints` here?

Reply:
I'm not all that familiar with the control flow where `ignoreConstraints` is set to true, but it guarded the insertion of valid constraints into the constraint set.

For constraints normalized to false (we know it is always unsat, eg Top <: Bottom), an error msg must be produced, so this seems like the best place to do it.

Reply:
Update: ignore all constraints if ignoreConstraint is set to true, including AlwaysFalseConstraint. (aka do nothing if flag = true)